src/app/message-reader.h

```diff
@@ -0,0 +1,138 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          Contains the API definition for a message buffer reader for the data
+ *          model.  This reader does the necessary bounds-checks before reading
+ *          and updates its own state as the buffer is read.
+ */
+
+#pragma once
+
+#include <app/util/basic-types.h>
+#include <lib/core/CHIPError.h>
+#include <lib/support/BufferReader.h>
+#include <lib/support/CodeUtils.h>
+#include <stdint.h>
+
+namespace chip {
+// Private inheritance because we don't necessarily want to expose all the
+// methods our superclass has.
+class DataModelReader : private Encoding::LittleEndian::Reader
+{
+    typedef Encoding::LittleEndian::Reader Super;
```

Comment:
Nit: Super sounds like an odd name. Maybe `BaseReader` ?

Is there a value in private derivation? I wonder if it would be easier to compose rather than derive, and then I would have:

```cpp
...
private:
   Encoding::LittleEndian::Reader mReader;
public:
   uint16_t OctetsRead() const { return mReader.OctetsRead(); }
    ...
```

This looks just as compact memory wise and it seems to avoid the extra typedef.

Reply:
Yeah, I can do the composition thing, good idea!

Reply:
I guess that requires me to make the templated `Read` public on `LittleEndian::Reader`.  I think I'm OK with that...

Reply:
Maybe call it 'RawRead' or something. May also be useful to directly write RawRead<uint8_t>(....) instead of just RawRead(...) to make if very explicit.

Or make this class conver ReadFoo into the correct Read16/Read32 and such. Tedious if someone decides to change sizes of some of our basic data types.

Reply:
Not "if", "when".  We know people want to change the sizes of some of them; I am trying to land the groundwork for making it painless....  Hence avoiding the explicit typing of `Read` or the use of specific `Read*` functions.